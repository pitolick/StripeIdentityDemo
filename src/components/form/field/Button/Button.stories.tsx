import type { Meta, StoryObj } from '@storybook/react'
import { Button } from './Button'

const meta: Meta<typeof Button> = {
  title: 'Form/Field/Button',
  component: Button,
  tags: ['autodocs'],
}

export default meta
type Story = StoryObj<typeof Button>

/**
 * 基本形
 */
export const Base: Story = {
  args: {
    children: 'Button',
  },
}

/**
 * タイプバリエーション：ボタン
 */
export const TypeButton: Story = {
  args: {
    children: 'Button',
    type: 'button',
  },
}

/**
 * タイプバリエーション：サブミット
 */
export const TypeSubmit: Story = {
  args: {
    children: 'Button',
    type: 'submit',
  },
}

/**
 * タイプバリエーション：リセット
 */
export const TypeReset: Story = {
  args: {
    children: 'Button',
    type: 'reset',
  },
}

/**
 * カラーバリエーション：黒（ベタ）
 */
export const ColorBlackSolid: Story = {
  args: {
    children: 'Button',
    color: 'black',
    buttonStyle: 'solid',
    iconLeft: 'arrow',
    iconRight: 'arrow',
  },
}

/**
 * カラーバリエーション：赤（ベタ）
 */
export const ColorRedSolid: Story = {
  args: {
    children: 'Button',
    color: 'red',
    buttonStyle: 'solid',
    iconLeft: 'arrow',
    iconRight: 'arrow',
  },
}

/**
 * カラーバリエーション：青（ベタ）
 */
export const ColorBlueSolid: Story = {
  args: {
    children: 'Button',
    color: 'blue',
    buttonStyle: 'solid',
    iconLeft: 'arrow',
    iconRight: 'arrow',
  },
}

/**
 * カラーバリエーション：グレー（ベタ）
 */
export const ColorGraySolid: Story = {
  args: {
    children: 'Button',
    color: 'gray',
    buttonStyle: 'solid',
    iconLeft: 'arrow',
    iconRight: 'arrow',
  },
}

/**
 * カラーバリエーション：黒（線）
 */
export const ColorBlackOutline: Story = {
  args: {
    children: 'Button',
    color: 'black',
    buttonStyle: 'outline',
    iconLeft: 'arrow',
    iconRight: 'arrow',
  },
}

/**
 * カラーバリエーション：赤（線）
 */
export const ColorRedOutline: Story = {
  args: {
    children: 'Button',
    color: 'red',
    buttonStyle: 'outline',
    iconLeft: 'arrow',
    iconRight: 'arrow',
  },
}

/**
 * カラーバリエーション：青（線）
 */
export const ColorBlueOutline: Story = {
  args: {
    children: 'Button',
    color: 'blue',
    buttonStyle: 'outline',
    iconLeft: 'arrow',
    iconRight: 'arrow',
  },
}

/**
 * カラーバリエーション：グレー（線）
 */
export const ColorGrayOutline: Story = {
  args: {
    children: 'Button',
    color: 'gray',
    buttonStyle: 'outline',
    iconLeft: 'arrow',
    iconRight: 'arrow',
  },
}

/**
 * 子要素にDOMを含める
 */
export const DOM_Children: Story = {
  args: {
    children: (
      <span className="text-center text-sm">
        Link
        <br />
        Button
      </span>
    ),
  },
}

/**
 * インラインボタン
 */
export const InlineButton: Story = {
  args: {
    children: 'Button',
    full: false,
  },
}

/**
 * インラインボタン（アイコンあり）
 */
export const InlineButton_Icon: Story = {
  args: {
    children: 'Button',
    full: false,
    iconLeft: 'arrow',
    iconRight: 'arrow',
  },
}

/**
 * アイコン（左側Arrow）
 */
export const IconLeftArrow: Story = {
  args: {
    children: 'Button',
    iconLeft: 'arrow',
  },
}

/**
 * アイコン（左側SVG）
 */
export const IconLeft: Story = {
  args: {
    children: 'Button',
    iconLeft: (
      <svg className="h-auto w-2.5 rotate-180" width="17" height="16" viewBox="0 0 17 16" fill="none" xmlns="http://www.w3.org/2000/svg">
        <path
          fillRule="evenodd"
          clipRule="evenodd"
          d="M1 7C0.447715 7 -3.73832e-07 7.44771 -3.49691e-07 8C-3.2555e-07 8.55228 0.447715 9 1 9L13.0858 9L7.79289 14.2929C7.40237 14.6834 7.40237 15.3166 7.79289 15.7071C8.18342 16.0976 8.81658 16.0976 9.20711 15.7071L16.0303 8.88388C16.5185 8.39573 16.5185 7.60427 16.0303 7.11612L9.20711 0.292893C8.81658 -0.0976318 8.18342 -0.0976318 7.79289 0.292893C7.40237 0.683417 7.40237 1.31658 7.79289 1.70711L13.0858 7L1 7Z"
          fill="currentColor"
        />
      </svg>
    ),
  },
}

/**
 * アイコン（右側Arrow）
 */
export const IconRightArrow: Story = {
  args: {
    children: 'Button',
    iconRight: 'arrow',
  },
}

/**
 * アイコン（右側SVG）
 */
export const IconRight: Story = {
  args: {
    children: 'Button',
    iconRight: (
      <svg className="h-auto w-2.5" width="17" height="16" viewBox="0 0 17 16" fill="none" xmlns="http://www.w3.org/2000/svg">
        <path
          fillRule="evenodd"
          clipRule="evenodd"
          d="M1 7C0.447715 7 -3.73832e-07 7.44771 -3.49691e-07 8C-3.2555e-07 8.55228 0.447715 9 1 9L13.0858 9L7.79289 14.2929C7.40237 14.6834 7.40237 15.3166 7.79289 15.7071C8.18342 16.0976 8.81658 16.0976 9.20711 15.7071L16.0303 8.88388C16.5185 8.39573 16.5185 7.60427 16.0303 7.11612L9.20711 0.292893C8.81658 -0.0976318 8.18342 -0.0976318 7.79289 0.292893C7.40237 0.683417 7.40237 1.31658 7.79289 1.70711L13.0858 7L1 7Z"
          fill="currentColor"
        />
      </svg>
    ),
  },
}
