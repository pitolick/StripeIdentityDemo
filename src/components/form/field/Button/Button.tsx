/**
 * ボタンコンポーネント
 * @module Button
 */

import React from 'react'

/**
 * ボタンコンポーネントのプロパティ
 * @typedef {Object} ButtonProps
 * @property {string} children - ボタンに表示するテキスト
 * @property {boolean} full - ブロックボタンにするか
 * @property {React.ButtonHTMLAttributes<HTMLButtonElement>} - ボタンタグのHTML属性
 */
interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  /**
   * 子要素
   */
  children: React.ReactNode
  /**
   * 色（デフォルトカラーはコンポーネント側で指定）
   */
  color?: 'black' | 'red' | 'blue' | 'gray'
  /**
   * ボタンのスタイル
   */
  buttonStyle?: 'solid' | 'outline'
  /**
   * ボタンのタイプ
   */
  type?: 'button' | 'submit' | 'reset'
  /**
   * trueでブロック要素化
   */
  full?: boolean
  /**
   * アイコンを設置（左側）
   * `arrow`で組み込みの矢印を表示
   */
  iconLeft?: React.ReactNode | 'arrow'
  /**
   * アイコンを設置（右側）
   * `arrow`で組み込みの矢印を表示
   */
  iconRight?: React.ReactNode | 'arrow'
}

/**
 * ボタンコンポーネント
 * @param {ButtonProps} props - ボタンコンポーネントのプロパティ
 * @returns {React.FC} - ボタンコンポーネント
 */
export const Button: React.FC<ButtonProps> = ({ children, type = 'button', color = 'blue', buttonStyle = 'solid', full = true, iconLeft, iconRight, ...buttonProps }) => {
  // 色
  let colorClass = ''
  let iconColorClass = ''
  switch (color) {
    case 'black':
      switch (buttonStyle) {
        case 'solid':
          colorClass = 'border-transparent text-white bg-black hover:bg-gray-900'
          iconColorClass = 'text-white'
          break
        case 'outline':
          colorClass = 'border-black text-black bg-white hover:text-white hover:bg-black'
          iconColorClass = 'text-black group-hover:text-white'
          break
      }
      break
    case 'red':
      switch (buttonStyle) {
        case 'solid':
          colorClass = 'border-transparent text-white bg-red hover:bg-red-light'
          iconColorClass = 'text-white'
          break
        case 'outline':
          colorClass = 'border-red text-red bg-white hover:text-white hover:bg-red'
          iconColorClass = 'text-red group-hover:text-white'
          break
      }
      break
    case 'blue':
      switch (buttonStyle) {
        case 'solid':
          colorClass = 'border-transparent text-white bg-blue hover:bg-blue-light'
          iconColorClass = 'text-white'
          break
        case 'outline':
          colorClass = 'border-blue text-blue bg-white hover:text-white hover:bg-blue'
          iconColorClass = 'text-blue group-hover:text-white'
          break
      }
      break
    case 'gray':
      switch (buttonStyle) {
        case 'solid':
          colorClass = 'border-transparent text-gray-900 bg-gray-500 hover:bg-gray-300'
          iconColorClass = 'text-gray-900'
          break
        case 'outline':
          colorClass = 'border-gray-500 text-gray-500 bg-white hover:border-transparent hover:text-gray-900 hover:bg-gray-500'
          iconColorClass = 'text-gray-500 group-hover:text-gray-900'
          break
      }
      break
  }

  // アイコン用の共通クラス
  const iconBaseClass = 'absolute inset-y-0 inline-flex items-center justify-center m-auto h-full w-3.5 '
  return (
    <button
      type={type}
      className={`
        ${full && 'w-full'} ${colorClass} group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      `}
      {...buttonProps}
    >
      {iconLeft && <span className={`${iconBaseClass} ${iconColorClass} left-4`}>{iconLeft}</span>}
      {children}
      {iconRight && <span className={`${iconBaseClass} ${iconColorClass} right-4`}>{iconRight}</span>}
    </button>
  )
}
