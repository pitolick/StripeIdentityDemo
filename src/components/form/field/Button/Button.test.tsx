import React from 'react'
import { render } from '@testing-library/react'
import { Button } from './Button'

describe('Buttonコンポーネント', () => {
  it('正しくレンダリングされること', () => {
    const { container } = render(<Button>テストボタン</Button>)
    expect(container).toMatchSnapshot()
  })

  it('fullプロパティがtrueの場合、ブロック要素としてレンダリングされること', () => {
    const { container } = render(<Button full>テストボタン</Button>)
    expect(container).toMatchSnapshot()
  })

  it('fullプロパティがfalseの場合、インライン要素としてレンダリングされること', () => {
    const { container } = render(<Button full={false}>テストボタン</Button>)
    expect(container).toMatchSnapshot()
  })

  it('colorプロパティが黒色でbuttonStyleプロパティがsolidを指定した場合、正しく適用されること', () => {
    const { container } = render(
      <Button color="black" buttonStyle="solid">
        テストボタン
      </Button>,
    )
    expect(container).toMatchSnapshot()
  })

  it('colorプロパティが黒色でbuttonStyleプロパティがoutlineを指定した場合、正しく適用されること', () => {
    const { container } = render(
      <Button color="black" buttonStyle="outline">
        テストボタン
      </Button>,
    )
    expect(container).toMatchSnapshot()
  })

  it('colorプロパティが赤色でbuttonStyleプロパティがsolidを指定した場合、正しく適用されること', () => {
    const { container } = render(
      <Button color="red" buttonStyle="solid">
        テストボタン
      </Button>,
    )
    expect(container).toMatchSnapshot()
  })

  it('colorプロパティが赤色でbuttonStyleプロパティがoutlineを指定した場合、正しく適用されること', () => {
    const { container } = render(
      <Button color="red" buttonStyle="outline">
        テストボタン
      </Button>,
    )
    expect(container).toMatchSnapshot()
  })

  it('colorプロパティが青色でbuttonStyleプロパティがsolidを指定した場合、正しく適用されること', () => {
    const { container } = render(
      <Button color="blue" buttonStyle="solid">
        テストボタン
      </Button>,
    )
    expect(container).toMatchSnapshot()
  })

  it('colorプロパティが青色でbuttonStyleプロパティがoutlineを指定した場合、正しく適用されること', () => {
    const { container } = render(
      <Button color="blue" buttonStyle="outline">
        テストボタン
      </Button>,
    )
    expect(container).toMatchSnapshot()
  })

  it('colorプロパティが灰色でbuttonStyleプロパティがsolidを指定した場合、正しく適用されること', () => {
    const { container } = render(
      <Button color="gray" buttonStyle="solid">
        テストボタン
      </Button>,
    )
    expect(container).toMatchSnapshot()
  })

  it('colorプロパティが灰色でbuttonStyleプロパティがoutlineを指定した場合、正しく適用されること', () => {
    const { container } = render(
      <Button color="gray" buttonStyle="outline">
        テストボタン
      </Button>,
    )
    expect(container).toMatchSnapshot()
  })

  it('iconLeftプロパティにarrowを指定した場合、正しく表示されること', () => {
    const { container } = render(<Button iconLeft="arrow">テストボタン</Button>)
    expect(container).toMatchSnapshot()
  })

  it('iconLeftプロパティにアイコンを指定した場合、正しく表示されること', () => {
    const { container } = render(<Button iconLeft={<span>アイコン</span>}>テストボタン</Button>)
    expect(container).toMatchSnapshot()
  })

  it('iconRightプロパティにarrowを指定した場合、正しく表示されること', () => {
    const { container } = render(<Button iconRight="arrow">テストボタン</Button>)
    expect(container).toMatchSnapshot()
  })

  it('iconRightプロパティにアイコンを指定した場合、正しく表示されること', () => {
    const { container } = render(<Button iconRight={<span>アイコン</span>}>テストボタン</Button>)
    expect(container).toMatchSnapshot()
  })
})
