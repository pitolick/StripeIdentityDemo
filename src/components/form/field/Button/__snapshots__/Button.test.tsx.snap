// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttonコンポーネント colorプロパティが灰色でbuttonStyleプロパティがoutlineを指定した場合、正しく適用されること 1`] = `
<div>
  <button
    class="
        w-full border-gray-500 text-gray-500 bg-white hover:border-transparent hover:text-gray-900 hover:bg-gray-500 group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    type="button"
  >
    テストボタン
  </button>
</div>
`;

exports[`Buttonコンポーネント colorプロパティが灰色でbuttonStyleプロパティがsolidを指定した場合、正しく適用されること 1`] = `
<div>
  <button
    class="
        w-full border-transparent text-gray-900 bg-gray-500 hover:bg-gray-300 group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    type="button"
  >
    テストボタン
  </button>
</div>
`;

exports[`Buttonコンポーネント colorプロパティが赤色でbuttonStyleプロパティがoutlineを指定した場合、正しく適用されること 1`] = `
<div>
  <button
    class="
        w-full border-red text-red bg-white hover:text-white hover:bg-red group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    type="button"
  >
    テストボタン
  </button>
</div>
`;

exports[`Buttonコンポーネント colorプロパティが赤色でbuttonStyleプロパティがsolidを指定した場合、正しく適用されること 1`] = `
<div>
  <button
    class="
        w-full border-transparent text-white bg-red hover:bg-red-light group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    type="button"
  >
    テストボタン
  </button>
</div>
`;

exports[`Buttonコンポーネント colorプロパティが青色でbuttonStyleプロパティがoutlineを指定した場合、正しく適用されること 1`] = `
<div>
  <button
    class="
        w-full border-blue text-blue bg-white hover:text-white hover:bg-blue group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    type="button"
  >
    テストボタン
  </button>
</div>
`;

exports[`Buttonコンポーネント colorプロパティが青色でbuttonStyleプロパティがsolidを指定した場合、正しく適用されること 1`] = `
<div>
  <button
    class="
        w-full border-transparent text-white bg-blue hover:bg-blue-light group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    type="button"
  >
    テストボタン
  </button>
</div>
`;

exports[`Buttonコンポーネント colorプロパティが黒色でbuttonStyleプロパティがoutlineを指定した場合、正しく適用されること 1`] = `
<div>
  <button
    class="
        w-full border-black text-black bg-white hover:text-white hover:bg-black group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    type="button"
  >
    テストボタン
  </button>
</div>
`;

exports[`Buttonコンポーネント colorプロパティが黒色でbuttonStyleプロパティがsolidを指定した場合、正しく適用されること 1`] = `
<div>
  <button
    class="
        w-full border-transparent text-white bg-black hover:bg-gray-900 group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    type="button"
  >
    テストボタン
  </button>
</div>
`;

exports[`Buttonコンポーネント fullプロパティがfalseの場合、インライン要素としてレンダリングされること 1`] = `
<div>
  <button
    class="
        false border-transparent text-white bg-blue hover:bg-blue-light group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    type="button"
  >
    テストボタン
  </button>
</div>
`;

exports[`Buttonコンポーネント fullプロパティがtrueの場合、ブロック要素としてレンダリングされること 1`] = `
<div>
  <button
    class="
        w-full border-transparent text-white bg-blue hover:bg-blue-light group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    type="button"
  >
    テストボタン
  </button>
</div>
`;

exports[`Buttonコンポーネント iconLeftプロパティにarrowを指定した場合、正しく表示されること 1`] = `
<div>
  <button
    class="
        w-full border-transparent text-white bg-blue hover:bg-blue-light group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    type="button"
  >
    <span
      class="absolute inset-y-0 inline-flex items-center justify-center m-auto h-full w-3.5  text-white left-4"
    >
      arrow
    </span>
    テストボタン
  </button>
</div>
`;

exports[`Buttonコンポーネント iconLeftプロパティにアイコンを指定した場合、正しく表示されること 1`] = `
<div>
  <button
    class="
        w-full border-transparent text-white bg-blue hover:bg-blue-light group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    type="button"
  >
    <span
      class="absolute inset-y-0 inline-flex items-center justify-center m-auto h-full w-3.5  text-white left-4"
    >
      <span>
        アイコン
      </span>
    </span>
    テストボタン
  </button>
</div>
`;

exports[`Buttonコンポーネント iconRightプロパティにarrowを指定した場合、正しく表示されること 1`] = `
<div>
  <button
    class="
        w-full border-transparent text-white bg-blue hover:bg-blue-light group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    type="button"
  >
    テストボタン
    <span
      class="absolute inset-y-0 inline-flex items-center justify-center m-auto h-full w-3.5  text-white right-4"
    >
      arrow
    </span>
  </button>
</div>
`;

exports[`Buttonコンポーネント iconRightプロパティにアイコンを指定した場合、正しく表示されること 1`] = `
<div>
  <button
    class="
        w-full border-transparent text-white bg-blue hover:bg-blue-light group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    type="button"
  >
    テストボタン
    <span
      class="absolute inset-y-0 inline-flex items-center justify-center m-auto h-full w-3.5  text-white right-4"
    >
      <span>
        アイコン
      </span>
    </span>
  </button>
</div>
`;

exports[`Buttonコンポーネント 正しくレンダリングされること 1`] = `
<div>
  <button
    class="
        w-full border-transparent text-white bg-blue hover:bg-blue-light group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    type="button"
  >
    テストボタン
  </button>
</div>
`;
